clang/lib/CIR/CodeGen/CIRGenExprAgg.cpp

```diff
@@ -1024,6 +1025,142 @@ void AggExprEmitter::withReturnValueSlot(
   }
 }
 
+enum CompareKind {
+  CK_Less,
+  CK_Greater,
+  CK_Equal,
+};
+
+static mlir::Value buildCompare(CIRGenFunction &CGF, const BinaryOperator *E,
+                                mlir::Value LHS, mlir::Value RHS,
+                                CompareKind Kind) {
+  QualType ArgTy = E->getLHS()->getType();
+  if (const ComplexType *CT = ArgTy->getAs<ComplexType>())
+    ArgTy = CT->getElementType();
+
+  if (const auto *MPT = ArgTy->getAs<MemberPointerType>())
+    llvm_unreachable("NYI");
+
+  auto Loc = CGF.getLoc(E->getSourceRange());
+
+  mlir::cir::CmpOpKind CmpKind;
+  switch (Kind) {
+  case CK_Less:
+    CmpKind = mlir::cir::CmpOpKind::lt;
+    break;
+
+  case CK_Greater:
+    CmpKind = mlir::cir::CmpOpKind::gt;
+    break;
+
+  case CK_Equal:
+    CmpKind = mlir::cir::CmpOpKind::eq;
+    break;
+
+  default:
+    llvm_unreachable("invalid compare kind");
+  }
+
+  return CGF.getBuilder().createCompare(Loc, CmpKind, LHS, RHS);
+}
+
+void AggExprEmitter::VisitBinCmp(const BinaryOperator *E) {
+  assert(CGF.getContext().hasSameType(E->getLHS()->getType(),
+                                      E->getRHS()->getType()));
+  const ComparisonCategoryInfo &CmpInfo =
+      CGF.getContext().CompCategories.getInfoForType(E->getType());
+  assert(CmpInfo.Record->isTriviallyCopyable() &&
+         "cannot copy non-trivially copyable aggregate");
+
+  QualType ArgTy = E->getLHS()->getType();
+
+  if (!ArgTy->isIntegralOrEnumerationType() && !ArgTy->isRealFloatingType() &&
+      !ArgTy->isNullPtrType() && !ArgTy->isPointerType() &&
+      !ArgTy->isMemberPointerType() && !ArgTy->isAnyComplexType()) {
+    return CGF.ErrorUnsupported(E, "aggregate three-way comparison");
+  }
+  bool IsComplex = ArgTy->isAnyComplexType();
+
+  auto EmitOperand = [&](Expr *E) -> std::pair<mlir::Value, mlir::Value> {
+    RValue RV = CGF.buildAnyExpr(E);
+    if (RV.isScalar())
+      return {RV.getScalarVal(), nullptr};
+    if (RV.isAggregate())
+      return {RV.getAggregatePointer(), nullptr};
```

Comment:
Is this path tested?

Reply:
I believe this path should never be executed. I have removed it.